### PR TITLE
🌱 add E2E test for MachineSet Preflight checks

### PR DESCRIPTION
**Pull request by ykakarap**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: test/e2e/cluster_upgrade_runtimesdk.go [hunk 1]
```diff
@@ -281,6 +286,71 @@ func clusterUpgradeWithRuntimeSDKSpec(ctx context.Context, inputGetter func() cl
...
+	// to the MachineDeployment. Therefore, the changes made to the MachineDeployment here will not be replaced
+	// immediately.
+	*md.Spec.Replicas += 1
+	Eventually(func() error {
+		return patchHelper.Patch(ctx, md)
+	}).Should(Succeed(), "Failed to scale up the MachineDeployment %s", klog.KObj(md))
+
+	// Verify that the MachineSet is not creating the new Machine.
+	Eventually(func(g Gomega) {
+		machineSets := framework.GetMachineSetsByCluster(ctx, framework.GetMachineSetsByClusterInput{
```
Comment (sbueringer) [inline test/e2e/cluster_upgrade_runtimesdk.go:334]: Wondering if we should get the 1 MS of the MD we scaled up? Currently this all relies on that we only have one MD and one MS in the cluster, which might be a bit brittle

**Comment by sbueringer** [comment]
@killianmuldoon When you have some time, can you do a sanity check on the last commit in this PR?

File: test/e2e/cluster_upgrade_runtimesdk_test.go [hunk 2]
```diff
@@ -26,7 +26,7 @@ import (
 	"k8s.io/utils/pointer"
 )
 
-var _ = Describe("When upgrading a workload cluster using ClusterClass with RuntimeSDK [PR-Informing] [ClusterClass]", func() {
+var _ = Describe("When upgrading a workload cluster using ClusterClass with RuntimeSDK [PR-Informing] [ClusterClass] [PR-Blocking]", func() {
```
Comment (ykakarap) [inline test/e2e/cluster_upgrade_runtimesdk_test.go:29]: This change is only to get quick test signal on the PR. The change will be dropped before merging.

Comment (sbueringer) [reply test/e2e/cluster_upgrade_runtimesdk_test.go:29]: Up to you, but in my opinion you can get rid of it already as the test is run in `pull-cluster-api-e2e-informing-main` anyway (which is also always triggered like e2e-main)

File: test/e2e/cluster_upgrade_runtimesdk.go [hunk 3]
```diff
@@ -281,6 +286,85 @@ func clusterUpgradeWithRuntimeSDKSpec(ctx context.Context, inputGetter func() cl
...
+			ClusterName: clusterRef.Name,
+			Namespace:   clusterRef.Namespace,
+		})
+		for i := range mds {
+			if mds[i].Name != md.Name {
+				continue
+			}
+			g.Expect(mds[i].Spec.Replicas).To(Equal(md.Spec.Replicas))
+		}
+	})
```
Comment (sbueringer) [inline test/e2e/cluster_upgrade_runtimesdk.go:344]: Let's set an explicit timeout on this consistently. I don't know what the default is and I don't want to rely on it.
I think 10 seconds would be fine

File: test/e2e/cluster_upgrade_runtimesdk.go [hunk 4]
```diff
@@ -281,6 +286,84 @@ func clusterUpgradeWithRuntimeSDKSpec(ctx context.Context, inputGetter func() cl
...
+		// Get the updated MachineDeployment.
+		targetMD := &clusterv1.MachineDeployment{}
+		// Wrap in an Eventually block for additional safety. Since all of this is in a Consistently block it
+		// will fail if we hit a transient error like a network flake.
+		Eventually(func() error {
+			return c.Get(ctx, client.ObjectKeyFromObject(md), targetMD)
+		}).Should(Succeed(), "Failed to get MachineDeployment %s", klog.KObj(md))
+		// Verify replicas are not overridden.
+		g.Expect(targetMD.Spec.Replicas).To(Equal(md.Spec.Replicas))
+	}, "10s", "1s")
```
Comment (sbueringer) [inline test/e2e/cluster_upgrade_runtimesdk.go:343]:
```suggestion
	}, 10*time.Second, 1*time.Second)
```
nit: let's use durations directly as we usually do (Just for consistency)

File: test/e2e/cluster_upgrade_runtimesdk.go [hunk 5]
```diff
@@ -281,6 +286,84 @@ func clusterUpgradeWithRuntimeSDKSpec(ctx context.Context, inputGetter func() cl
...
+		return patchHelper.Patch(ctx, md)
+	}).Should(Succeed(), "Failed to scale up the MachineDeployment %s", klog.KObj(md))
+	// Verify the MachineDeployment updated replicas are not overridden.
+	// Note: This verifies that the topology controller in fact holds any reconciliation of this MachineDeployment.
+	Consistently(func(g Gomega) {
+		// Get the updated MachineDeployment.
+		targetMD := &clusterv1.MachineDeployment{}
+		// Wrap in an Eventually block for additional safety. Since all of this is in a Consistently block it
+		// will fail if we hit a transient error like a network flake.
+		Eventually(func() error {
```
Comment (sbueringer) [inline test/e2e/cluster_upgrade_runtimesdk.go:338]:
```suggestion
		g.Eventually(func() error {
```
Not entirely sure but probably better to use the same gomega

**Review by sbueringer — COMMENTED** [review]
Last round of nits from my side

**Comment by sbueringer** [comment]
Lgtm pending underlying PRs are merged + squash/rebase

**Comment by jackfrancis** [comment]
This PR looks like it includes actual implementation of these checks and not just E2E, should we retitle?

**Comment by sbueringer** [comment]
Those commits will be dropped. We have to merge underlying PRs first which is also the reason why the PR is on hold. xref https://github.com/kubernetes-sigs/cluster-api/pull/8698#issuecomment-1574453784

File: test/e2e/cluster_upgrade_runtimesdk.go [hunk 6]
```diff
@@ -281,6 +286,84 @@ func clusterUpgradeWithRuntimeSDKSpec(ctx context.Context, inputGetter func() cl
...
 }
 
+// machineSetPreflightChecksTestHandler verifies the MachineSet preflight checks.
+// At this point in the test the ControlPlane is upgraded to the new version and the upgrade to the MachineDeployments
+// should be blocked by the AfterControlPlaneUpgrade hook.
+// Test the MachineSet preflight checks by scaling up the MachineDeployment. The creation on the new Machine
+// should be blocked because the preflight checks should not pass (kubeadm version skew preflight check should fail).
+func machineSetPreflightChecksTestHandler(ctx context.Context, c client.Client, clusterRef types.NamespacedName) {
+	// Verify that the hook is called and the topology reconciliation is blocked.
+	hookName := "AfterControlPlaneUpgrade"
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply test/e2e/cluster_upgrade_runtimesdk.go:296]: It is only called if there is a Runtime Extension deployed. This test is the only one who uses Runtime Extensions.

**Comment by sbueringer** [comment]
@ykakarap Just merged the underlying PR, can you please rebase this PR?

**Comment by ykakarap** [comment]
> @ykakarap Just merged the underlying PR, can you please rebase this PR?

Will do.

**Comment by ykakarap** [comment]
Rebased.

File: test/e2e/cluster_upgrade_runtimesdk.go [hunk 7]
```diff
@@ -281,6 +286,84 @@ func clusterUpgradeWithRuntimeSDKSpec(ctx context.Context, inputGetter func() cl
...
+
+	// Scale up the MachineDeployment
+	machineDeployments := framework.GetMachineDeploymentsByCluster(ctx, framework.GetMachineDeploymentsByClusterInput{
+		Lister:      c,
+		ClusterName: clusterRef.Name,
+		Namespace:   clusterRef.Namespace,
+	})
+	md := machineDeployments[0]
+
+	patchHelper, err := patch.NewHelper(md, c)
```
Comment (fabriziopandini) [inline test/e2e/cluster_upgrade_runtimesdk.go:320]: If I'm not wrong this E2E test is using a cluster class, so we should not patch the cluster and not the md..

Comment (ykakarap) [reply test/e2e/cluster_upgrade_runtimesdk.go:322]: Discussed with Fabrizio and added some more comments to explain more about the test and the way the test is performed. 

Additionally added another check to make sure no new machines are created when a MachineSet preflight check is blocking new machine creation.

**Pull request merged**